src/ipc/ipc4/helper.c

```diff
@@ -401,13 +406,20 @@ int ipc_comp_connect(struct ipc *ipc, ipc_pipe_comp_connect *_connect)
 	}
 
 	/* both sink and source components should have their direction set during module binding */
-	if (!sink->direction_set || !source->direction_set)
-		return IPC4_INVALID_RESOURCE_STATE;
+	if (!sink->direction_set || !source->direction_set) {
+		comp_unbind(source, bu);
+		comp_unbind(sink, bu);
+		ret = IPC4_INVALID_RESOURCE_STATE;
+		goto free;
+	}
 
 	return IPC4_SUCCESS;
 
 err:
 	irq_local_enable(flags);
+free:
+	pipeline_disconnect(source, buffer, PPL_CONN_DIR_COMP_TO_BUFFER);
+	pipeline_disconnect(sink, buffer, PPL_CONN_DIR_BUFFER_TO_COMP);
 	buffer_free(buffer);
```

Comment:
can we make it
```
e_sink:
	comp_unbind(sink, bu);
	pipeline_disconnect(sink, buffer, PPL_CONN_DIR_BUFFER_TO_COMP);
e_source:
	comp_unbind(source, bu);
	pipeline_disconnect(source, buffer, PPL_CONN_DIR_BUFFER_TO_COMP);
free:
	buffer_free(buffer);
```
and then jump to the correct label above?

Reply:
@lyakh I can't quite do that but i've squashed the 2 commits now to perform both connect and bind with the IRQ disabled and jump to the right label in case of errors. Let me know how it looks